drivers: Add MDIO bit-bang driver

meshium opened the pull request:
Add MDIO driver that works through GPIO pins. This driver is useful when a microcontroller doesn't have MDIO bus or when multiple separate MDIO buses are required.

meshium commented:
Hi All.
Could you please review my PR?

MaureenHelm commented:
@tbursztyka @rlubos please take a look

nashif commented:
@arvinf  Can you please take a look?

nashif commented:
@meshium  who is using this driver? we need a user (board, soc), ideally a build test needs to be added to tests/drivers/build_all

meshium commented:
@nashif This driver is not for any specific platform. The driver provides access to the MDIO bus through GPIO pins for any SoC that has GPIO pin control available. In particular, I have tested the driver's functionality on the Nucleo H743Zi board. If necessary, I can add a build test.

soburi reviewed (commented):
For now, it is pointing out just a matter of coding style.

cfriedt commented:
Please split adding the shell utility to a separate commit.

Please add the driver to a build_all testsuite.

cfriedt reviewed (changes requested):
Needs to be part of at least a build_all testsuite

nandojve commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,189 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mdio_gpio, CONFIG_MDIO_LOG_LEVEL);
+
+#define _READ  0
```
Just a remind about prefix. Usually what we do is **_MDIO_GPIO_**_READ. This avoid to clash with any already defined symbol which can came from includes.

Consider adding this for all macros that you defined here.

meshium replied:
Done

soburi replied:
Off-topic:
Please don't push 'Resolve conversation' button.
Reviewers are usually forget tomorrow 'What I pointed'.
If the button press and mark resolved, It make reviewer can't remind it.
So, the button should push by reviewer.

meshium commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,189 @@
+#define _READ  0
+#define _WRITE 1
+
+#define MDC_CLOCK_THE_BIT                                                                          \
```
Done

soburi replied:
It seems not fix yet.
Same also need to fix MDIO_GPIO_DIR, MDIO_GPIO_READ and MDIO_GPIO_WRITE.

meshium replied:
Sorry, I misunderstood your remark.
But have you noticed that I use local dev_cfg in those macros? I think readability will worsen if I'll rewrite them with ALWAYS_INLINE

soburi replied:
I think it is short enough for if changed. (not reach 80-chars width!)
It is good to define a macro that only pass the config argument to inline function If you really want abbreviate the argument.

Such like following...
```
#define MDIO_GPIO_READ(pdata) mdio_gpio_read(dev_cfg, pdata)
```

For syntax checking and debugging purposes, it is preferable to not use macros and let the compiler process them.

meshium replied:
Ok, it's resonable. Please take a look to the new commit.

meshium commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,189 @@
+		uint32_t v_len = (len);                                                            \
+		v_data <<= (32 - v_len);                                                           \
+		for (; v_len > 0; v_len--) {                                                       \
+			gpio_pin_set_dt(&dev_cfg->mdio_gpio, (v_data & 0x80000000) ? 1 : 0);       \
```
Done

meshium commented:
>

meshium commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,189 @@
+	struct gpio_dt_spec mdio_gpio;
+};
+
+static int mdio_transfer(const struct device *dev, uint8_t prtad, uint8_t devad, uint8_t rw,
```
Done

meshium commented:
@nandojve @cfriedt Can you please take a look?

soburi commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+
+static ALWAYS_INLINE void mdio_gpio_clock_the_bit(dev_cfg_ptr dev_cfg)
+{
+
```
nitpick: Remove this empty line.

meshium replied:
Fixed

soburi commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+
+static ALWAYS_INLINE void mdio_gpio_dir(dev_cfg_ptr dev_cfg, uint8_t dir)
+{
+
```
nitpick: Remove this empty line.

meshium replied:
Fixed

soburi commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+		if (gpio_pin_get_dt(&dev_cfg->mdio_gpio) == 1) {
+			data |= 1;
+		}
+	}
```
nitpick: Insert blank line.

meshium replied:
Fixed

soburi commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+		return rc;
+	}
+
+	if (!device_is_ready(dev_cfg->mdio_gpio.port)) {
```
This check should better to place before previous `before gpio_pin_configure_dt`.
It can reduce modification possibility of gpio status even if errors.

meshium replied:
Fixed

soburi commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+	struct gpio_dt_spec mdio_gpio;
+};
+
+typedef const struct mdio_gpio_config *dev_cfg_ptr;
```
Don't define alias.
Declare type name directly.
We don't prefer define alias to abbreviate in zephyr.

soburi commented:
Still remains CI error. Fix please.

nandojve commented:
Hi @meshium ,

Please fix compliance check.

BTW, I dropped `DNM` because I think it didn't apply anymore.

soburi commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+/*
```
Squash all commits that fix what the review points.

cfriedt replied:
Probably outdated?

soburi reviewed (changes requested):
Squash all commits that fix what the review points.

cfriedt reviewed (changes requested):
Can we please have >= 2 separate commits with commit messages that follow the [Commit Message Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines)?

Typically, we see a code commit first, and then a commit to tests.
E.g.
* drivers: mdio: add bit-bang driver
* tests: drivers: build_all: add mdio testsuite (if you wanted to have cleaner git history)
* tests: drivers: build_all: mdio: add bit bang testsuite

meshium committed:
drivers: mdio: add bit-bang driver

Add MDIO driver that works through GPIO pins. This driver is
useful when a microcontroller doesn't have MDIO bus or when multiple
separate MDIO buses are required. The driver provides access to
the MDIO bus through GPIO pins for any SoC that has GPIO pin control
available.

Signed-off-by: Aleksandr Senin <al@meshium.net>

meshium committed:
tests: drivers: build_all: mdio: add bit bang testsuite

- Add mdio to a build_all testsuite
- Add mdio bit bang to a build_all testsuite

Signed-off-by: Aleksandr Senin <al@meshium.net>

meshium commented:
@cfriedt @soburi @nandojve Do you have any further feedback on the review?

cfriedt commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+static ALWAYS_INLINE void mdio_gpio_dir(const struct mdio_gpio_config *dev_cfg, uint8_t dir)
+{
+	gpio_pin_configure_dt(&dev_cfg->mdio_gpio, (dir) ? GPIO_OUTPUT_ACTIVE : GPIO_INPUT);
+	if ((dir) == 0) {
```
```suggestion
	if (dir == 0) {
```

cfriedt commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+static ALWAYS_INLINE void mdio_gpio_write(const struct mdio_gpio_config *dev_cfg,
+					  uint32_t data, uint8_t len)
+{
+	uint32_t v_data = (data);
+	uint32_t v_len = (len);
```
```suggestion
	uint32_t v_data = data;
	uint32_t v_len = len;
```

cfriedt commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+	uint32_t v_data = (data);
+	uint32_t v_len = (len);
+
+	v_data <<= (32 - v_len);
```
```suggestion
	v_data <<= 32 - v_len;
```

cfriedt reviewed (commented):
Looks good to me, aside from a couple of nits.

soburi reviewed (dismissed):
No addition to say from me. Please fix @cfriedt pointings.

cfriedt reviewed (changes requested):
-1 until nits are fixed.

meshium commented:
> -1 until nits are fixed.

Done

meshium commented on drivers/mdio/mdio_gpio.c:
```diff
@@ -0,0 +1,195 @@
+
+static ALWAYS_INLINE void mdio_gpio_dir(const struct mdio_gpio_config *dev_cfg, uint8_t dir)
+{
+	gpio_pin_configure_dt(&dev_cfg->mdio_gpio, (dir) ? GPIO_OUTPUT_ACTIVE : GPIO_INPUT);
```
@cfriedt it's ok. fixed

meshium commented:
@nandojve Do you have any further feedback on the review?

meshium commented:
Hi @cfriedt, @nandojve, @soburi! What's next?
